File: analysis/tendencias_anuais.py
from typing import Any

import pandas as pd
from sqlalchemy import text

from analysis import fontes_receita


def _sum_col(conn: Any, table: str, col: str, year: int, root_only: bool = False) -> float:
    try:
        sql = f"SELECT {col} FROM {table} t WHERE t.ano = :ano"
        if root_only:
            sql += (
                f" AND NOT EXISTS ("
                f"SELECT 1 FROM {table} t2"
                f" WHERE t2.ano = :ano"
                f" AND t2.codigo != t.codigo"
                f" AND t.codigo LIKE RTRIM(t2.codigo, '0.') || '%%'"
                f" AND LENGTH(RTRIM(t2.codigo, '0.')) < LENGTH(RTRIM(t.codigo, '0.'))"
                f")"
            )
        df = pd.read_sql_query(text(sql), conn, params={"ano": year})
        if df.empty:
            return 0.0
        return float(pd.to_numeric(df[col].astype(str).str.replace(",", "."), errors="coerce").fillna(0).sum())
    except Exception:
        return 0.0


def run(conn: Any, years: list[int]) -> pd.DataFrame:
    records = []
    sorted_years = sorted(years)
    # Chama o módulo unificado de fontes de receita para obter as receitas de forma consolidada e DRY
    rev_df = fontes_receita.run(conn, sorted_years)
    rev_map = dict(zip(rev_df["ano"], rev_df["total_arrecadado"]))

    for year in sorted_years:
        total_gasto = _sum_col(conn, "despesas_por_orgao", "pago", year)
        total_folha = _sum_col(conn, "pessoal", "proventos", year)

        total_rec = rev_map.get(year, 0.0)
        receita = total_rec if total_rec > 0 else None
        restos = _sum_col(conn, "despesas_restos_pagar", "pago", year)

        records.append(
            {
                "ano": year,
                "total_gasto": total_gasto,
                "total_folha": total_folha,
                "total_receita": receita,
                "restos_a_pagar": restos,
            }
        )

    df = pd.DataFrame(records)
    for col in ["total_gasto", "total_folha", "total_receita", "restos_a_pagar"]:
        df[f"{col}_pct_change"] = df[col].pct_change() * 100
    return df


def gap_pressao_fiscal(df: pd.DataFrame) -> dict:
    """Retorna gap entre crescimento dos gastos e da receita, pronto para renderização em gráfico."""
    valid = df.dropna(subset=["total_gasto_pct_change"]).copy()
    valid = valid[~valid["total_gasto_pct_change"].isin([float("inf"), float("-inf")])]
    valid = valid.dropna(subset=["total_gasto_pct_change"])
    gap = [
        round(v, 2) for v in (valid["total_gasto_pct_change"] - valid["total_receita_pct_change"].fillna(0)).tolist()
    ]
    return {
        "anos": valid["ano"].tolist(),
        "gap": gap,
        "colors": ["#F44336" if v > 0 else "#4CAF50" for v in gap],
    }

File: packages/db/src/queries/tendencias_anuais.ts
import { sql } from "kysely";
import { db } from "../client";
import { getFontesReceita } from "./fontes_receita";

export interface TendenciaAnualRecord {
  ano: number;
  totalGasto: number;
  totalEmpenhado: number;
  totalFolha: number;
  totalReceita: number | null;
  restosAPagar: number;
  totalGastoPctChange?: number | null;
  totalEmpenhadoPctChange?: number | null;
  totalFolhaPctChange?: number | null;
  totalReceitaPctChange?: number | null;
  restosAPagarPctChange?: number | null;
}

export async function getTendenciasAnuais(
  years: number[],
  empresaIds?: string[] | null,
  portalSlug: string = "porciuncula_prefeitura",
): Promise<TendenciaAnualRecord[]> {
  const sortedYears = [...years].sort((a, b) => a - b);
  const revDf = await getFontesReceita(sortedYears, empresaIds);
  const revMap = new Map<number, number>();
  for (const r of revDf) {
    revMap.set(r.ano, r.totalArrecadado);
  }

  const records: TendenciaAnualRecord[] = [];

  for (const year of sortedYears) {
    let totalGasto = 0;
    let totalEmpenhado = 0;
    try {
      let q = sql`SELECT pago, empenhado FROM fct_despesas_por_orgao WHERE ano = ${year}`;
      if (empresaIds && empresaIds.length > 0) {
        q = sql`${q} AND empresa = ANY(${empresaIds})`;
      }
      const res = await q.execute(db);
      for (const r of (res.rows as any[]) || []) {
        totalGasto += parseFloat(String(r.pago ?? "0").replace(",", ".")) || 0;
        totalEmpenhado +=
          parseFloat(String(r.empenhado ?? "0").replace(",", ".")) || 0;
      }
    } catch {}

    let totalFolha = 0;
    try {
      const resF =
        await sql`SELECT proventos FROM fct_pessoal WHERE ano = ${year} AND portal_slug = ${portalSlug}`.execute(
          db,
        );
      for (const r of (resF.rows as any[]) || []) {
        totalFolha +=
          parseFloat(String(r.proventos ?? "0").replace(",", ".")) || 0;
      }
    } catch {}

    const totalRec = revMap.get(year) || 0;
    const receita = totalRec > 0 ? totalRec : null;

    let restos = 0;
    try {
      let qR = sql`SELECT pago FROM fct_despesas WHERE ano = ${year} AND fonte = 'restos_a_pagar' AND portal_slug = ${portalSlug}`;
      if (empresaIds && empresaIds.length > 0) {
        qR = sql`${qR} AND empresa_id = ANY(${empresaIds})`;
      }
      const resR = await qR.execute(db);
      for (const r of (resR.rows as any[]) || []) {
        restos += parseFloat(String(r.pago ?? "0").replace(",", ".")) || 0;
      }
    } catch {}

    records.push({
      ano: year,
      totalGasto,
      totalEmpenhado,
      totalFolha,
      totalReceita: receita,
      restosAPagar: restos,
    });
  }

  for (let i = 0; i < records.length; i++) {
    if (i === 0) {
      records[i].totalGastoPctChange = null;
      records[i].totalEmpenhadoPctChange = null;
      records[i].totalFolhaPctChange = null;
      records[i].totalReceitaPctChange = null;
      records[i].restosAPagarPctChange = null;
    } else {
      const prev = records[i - 1];
      const curr = records[i];

      curr.totalGastoPctChange =
        prev.totalGasto > 0
          ? ((curr.totalGasto - prev.totalGasto) / prev.totalGasto) * 100
          : null;
      curr.totalEmpenhadoPctChange =
        prev.totalEmpenhado > 0
          ? ((curr.totalEmpenhado - prev.totalEmpenhado) /
              prev.totalEmpenhado) *
            100
          : null;
      curr.totalFolhaPctChange =
        prev.totalFolha > 0
          ? ((curr.totalFolha - prev.totalFolha) / prev.totalFolha) * 100
          : null;
      curr.totalReceitaPctChange =
        prev.totalReceita && prev.totalReceita > 0 && curr.totalReceita
          ? ((curr.totalReceita - prev.totalReceita) / prev.totalReceita) * 100
          : null;
      curr.restosAPagarPctChange =
        prev.restosAPagar > 0
          ? ((curr.restosAPagar - prev.restosAPagar) / prev.restosAPagar) * 100
          : null;
    }
  }

  return records;
}
